builtin/common/misc_helpers.lua

```diff
@@ -700,3 +714,68 @@ function core.privs_to_string(privs, delim)
 	end
 	return table.concat(list, delim)
 end
+
+--[[ Helper function for parsing an optionally relative number
+of a chat command parameter, using the chat command tilde notation.
+
+Parameters:
+* arg: String snippet containing the number; possible values:
+    * "<number>": return as number
+    * "~<number>": return absvalue + <number>
+    * "~": return absvalue
+    * Anything else will return `nil`
+* absvalue: Number to which the `arg` number might be relative to
+
+Returns:
+A number or `nil`, depending on `arg.
+
+Examples:
+* `core.parse_relative_number("5", 10)` returns 5
+* `core.parse_relative_number("~5", 10)` returns 15
+* `core.parse_relative_number("~", 10)` returns 10
+]]
+function core.parse_relative_number(arg, absvalue)
+	if not arg then
+		return nil
+	elseif arg == "~" then
+		return absvalue
+	elseif string.sub(arg, 1, 1) == "~" then
+		local number = tonumber(string.sub(arg, 2))
+		if not number then
+			return nil
+		end
+		return absvalue + number
+	else
+		return tonumber(arg)
+	end
+end
+
+--[[ Helper function to parse coordinates that might be relative
+to a player; supports chat command tilde notation.
+Intended to be used in chat command parameter parsing.
+
+Parameters:
+* x, y, z: Parsed x, y, and z coordinates as strings
+* player_name: Name of player of which to compare the position
+
+Syntax of x, y and z:
+* "<number>": return as number
+* "~<number>": return <number> + player position on this axis
+* "~": return player position on this axis
+
+Returns: a vector or nil for invalid input or if player does not exist
+]]
+function core.parse_coordinates(x, y, z, player_name)
+	local player = core.get_player_by_name(player_name)
+	local absolute_position = player:get_pos()
+	if not absolute_position then
+		return nil
+	end
+	local rx = core.parse_relative_number(x, absolute_position.x)
+	local ry = core.parse_relative_number(y, absolute_position.y)
+	local rz = core.parse_relative_number(z, absolute_position.z)
+	if not rx or not ry or not rz then
+		return nil
+	end
+	return { x = rx, y = ry, z = rz }
```

Comment:
How about
```lua
function core.parse_coordinates(coords, player_name)
	local player = core.get_player_by_name(player_name)
	local absolute_position = player:get_pos()
	if not absolute_position then
		return nil
	end
        local total_coords = {}
        for coord, value in pairs(coords) do
              local total_axis = core.parse_relative_number(value, absolute_position[coord])
              if not total_axis then
                  return nil
               end
        end
	return total_coords
end```

Reply:
Line count is almost identical, and performance too. For three indices only there's no need to loop.